src/System.Management.Automation/engine/InternalCommands.cs

```diff
@@ -16,6 +18,38 @@
 
 namespace Microsoft.PowerShell.Commands
 {
+    /// <summary>
+    /// A thin wrapper over a property-getting Callsite, to allow reuse when possible.
+    /// </summary>
+    class DynamicPropertyGetter
+    {
+        private CallSite<Func<CallSite, object, object>> _getValueDynamicSite;
+
+        // For the wildcard case, lets us know if we can reuse the callsite:
+        private string _lastUsedPropertyName;
+
+        public object GetValue(PSObject inputObject, string propertyName)
+        {
+            Dbg.Assert(!WildcardPattern.ContainsWildcardCharacters(propertyName), "propertyName should be pre-resolved by caller");
+
+            // If wildcards are involved, the resolved property name could potentially
+            // be different on every object... but probably not, so we'll attempt to
```

Comment:
I don't have any intuition for how often wildcards are used to specify multiple properties, but I know *I* use them that way for formatting commands.

I'm not sure perf matters much here, but it would be good to understand the impact on something like `ps | select *mem*`.

Reply:
For `select`, the wildcard gets resolved into individual `MshExpressionResult` objects, so no problems there. So this case here really is for the heterogenous-objects-with-different-matching-names case, which I think is probably a rare enough case not to care too much about (though it will work, of course).